test/io/formatters/Json.chpl

```diff
@@ -0,0 +1,231 @@
+      reader.readLiteral(":");
+      m[name] = reader.offset();
+      reader.revert();
+
+      qio_channel_skip_json_field(false, reader._channel_internal);
+      reader.matchLiteral(",");
+    }
+
+    reader.revert();
+
+    return (m, lastPos);
+  }
+
+  record JsonReader {
+    var outOfOrder = false;
+    var _inheritLevel = 0;
+    var _names : domain(string);
+    var _offsets : [_names] int;
+    var _lastPos = -1;
+
+    proc init() {
+      this.complete();
+    }
+
+    proc decode(reader:fileReader, type readType) : readType throws {
+      // TODO:
+      // - escaped strings
+      // - arrays/lists
+      if isNumericType(readType) || isBoolType(readType) {
+        var x : readType;
+        reader._readOne(reader.kind, x, here);
+        return x;
+      } else if isStringType(readType) {
+        var tmp : readType;
+        reader.readf("%{\"S\"}", tmp);
+        return tmp;
+      } else if isEnumType(readType) {
+        reader.readLiteral('"');
+        var ret = reader.withFormatter(DefaultReader).read(readType);
+        reader.readLiteral('"');
+        return ret;
+      } else if isClassType(readType) {
+        if isNilableClassType(readType) && reader.matchLiteral("null") {
+          return nil:readType;
+        } else {
+          return new readType(reader.withFormatter(new JsonReader()));
+        }
+      } else if canResolveTypeMethod(readType, "decodeFrom", reader) {
+        return readType.decodeFrom(reader.withFormatter(new JsonReader()));
+      } else {
+        return new readType(reader.withFormatter(new JsonReader()));
+      }
+    }
+
+    proc _readFieldName(r: _readerT, key: string) throws {
+      try {
+        r._readLiteral('"');
+        r._readLiteral(key);
+        r._readLiteral('"');
+        r._readLiteral(":");
+      } catch e: BadFormatError {
+        return false;
+      }
+
+      return true;
+    }
+
+
+    proc readField(r: _readerT, key: string, type T) throws {
+      if _names.contains(key) {
+        r.seek(_offsets[key]..);
+      } else if !key.isEmpty() {
+        throw new Error("field not found...");
+      }
+
+      var ret = r.read(T);
+
+      // note: trailing commas not allowed in json
+      r.matchLiteral(",");
+
+      return ret;
+    }
+
+    proc decodeField(r: _readerT, key: string, type T) throws {
+      return readField(r, key, T);
+    }
+
+    proc readTypeStart(r: fileReader, type T) throws {
+      if _inheritLevel == 0 {
+        if isClassType(T) || isRecordType(T) {
+          //
+          // TODO: When should we try to do this? Use of 'readTypeStart'
+          // likely indicates "buying in" to the JSON format, but in a way
+          // it might be cleaner or more efficient to do when creating the
+          // new temporary JsonReader for the given type. But if we do the
+          // JSON parsing at that point, it ignores the user-defined
+          // initializer. So, we do it here in 'readTypeStart'.
```

Comment:
Another consideration: ideally, it would only compute the field-to-offset mapping if the read is out of order.